### DISCUSSION TITLE
Private github repo support?

### DISCUSSION THREAD

**Discussion by casperthulebd** [microsoft/winget-cli#3151] [Q&A]
I would like to create a private github repository that hosts some manifests.
According to https://www.reddit.com/r/sysadmin/comments/if6it1/winget_local_repo/ it should be possible.

I have tried to create a private repository and copied the eclipse adoptium jdk 8 manifests file in there, similar to the structure here:
https://github.com/microsoft/winget-pkgs/tree/master/manifests

The question is now - how do I add this repository to winget?
I have tried 
`Winget source add --name test https://github.com/<OWNER>/<REPO>/tree/master/manifests --header "Authorization: token <TOKEN>"`

However, the response is:

    Ignoring the optional header as it is not applicable for this source.
    Adding source:
      test -> https://github.com/<OWNER>/REPO/tree/master/manifests
    An unexpected error occurred while executing the command:
    0x80190194 : Not found (404).
    
I am guessing that either I am missing something completely, such as:
I need to do something similar to https://github.com/cloudflightio/winget-pkgs/tree/main/.github/workflows
or the URL is wrong.

**Comment by Trenly** [comment]
> I would like to create a private github repository that hosts some manifests. According to [reddit.com/r/sysadmin/comments/if6it1/winget_local_repo](https://www.reddit.com/r/sysadmin/comments/if6it1/winget_local_repo/) it should be possible.
> 
> I have tried to create a private repository and copied the eclipse adoptium jdk 8 manifests file in there, similar to the structure here: [microsoft/winget-pkgs@`master`/manifests](https://github.com/microsoft/winget-pkgs/tree/master/manifests)
> 
> The question is now - how do I add this repository to winget? I have tried `Winget source add --name test https://github.com/<OWNER>/<REPO>/tree/master/manifests --header "Authorization: token <TOKEN>"`
> 
> However, the response is:
> 
> ```
> Ignoring the optional header as it is not applicable for this source.
> Adding source:
>   test -> https://github.com/<OWNER>/REPO/tree/master/manifests
> An unexpected error occurred while executing the command:
> 0x80190194 : Not found (404).
> ```
> 
> I am guessing that either I am missing something completely, such as: I need to do something similar to [cloudflightio/winget-pkgs@`main`/.github/workflows](https://github.com/cloudflightio/winget-pkgs/tree/main/.github/workflows) or the URL is wrong.

The Winget-Pkgs repository has an azure pipeline that compiles the manifests into a Pre-Indexed package. It's the msix package that is created which is used as the source. The other option would be to set up a Rest API to serve the manifest contents. I'm not familiar with how to set up either a Pre-Indexed source or a Rest source, but perhaps [microsoft/winget-cli-restsourcei](/microsoft/winget-cli-restsource) has documentation that may help

**Reply by casperthulebd** [reply]
Ahh, this was the missing piece:
https://stackoverflow.com/questions/61916158/structure-of-winget-source-repositories

It is possible to deploy a msix and thereby avoid having the rest api.

I am still a bit unsure of whether it has to be hosted as github pages as cloudflightio does or it can be directly in the repository. If directly in the repository, then how do I reference it? Through https://api.github.com/repos/<OWNER>/<REPO>/contents/<PATH_TO_FOLDER_WITH_SOURCE_MSIX> with 
--header 'Accept: application/vnd.github.v3.raw'?

**Reply by denelon** [reply]
The REST API is the only documented method for standing up a private repository.

**Comment by mherrmann** [comment]
You can try https://winget.Pro. It's a private repository implementation that can be hosted for you, or deployed in your own infrastructure via its open source version. (I'm one of the authors of this service.)

**Comment by vedantmgoyal9** [comment]
@casperthulebd you can see https://github.com/PckgrBot/winget-pkgs-private. It uses Vercel for APIs and https://neon.tech for storing metadata about packages. I'll add documentation soon, but until then I'll be happy to to help you set it up.

**Comment by Karl-WE** [comment]
@denelon what's the best method / docs using winget with either

- on local SMB fileshare 
- Azure Files

This is becoming a FAQ from customers, which do not only rely on Intune but want to embrace the upcoming Windows Server integration.

Looking for a solution on this and what I have found is anything but straight forward. Noted winget.pro and other paid solution, but LOB I would need are still missing and they want to host it on their own without additional licensing costs.

**Reply by mherrmann** [reply]
winget.pro is [open source](https://github.com/omaha-consulting/winget.pro) so you absolutely can host it on your own without additional licensing costs.

**Comment by Karl-WE** [comment]
@denelon what's the best method / docs using winget with either

- on local SMB fileshare 
- Azure Files

This is becoming a FAQ from customers, which do not only rely on Intune but want to embrace the upcoming Windows Server integration.

Looking for a solution on this and what I have found is anything but straight forward. I noticed winget.pro but LOB I would need are still missing.

Wouldn't be too bad if there was a "one-click" deployment of such a private repository available from Azure Marketplace that puts everything together via bicep or sorts.

remote related #619

**Reply by denelon** [reply]
@Karl-WE we're looking into an Azure Marketplace offer for the WinGet source to make this exact scenario easier to deploy. We've also got work to do for making a source easier to manage. We need to add a REST source as a target for wingetcreate, and we're looking into making a Microsoft.WinGet.Source PowerShell module so one could "Export-WinGetPackage" and subsequently "Publish-WinGetPackage". I'm also thinking we will want some kind of a "Modify" cmdlet so the installer could be copied to an enterprise managed location and the manifest could be modified to point to that URL.

**Comment by Karl-WE** [comment]
@mherrmann thanks for the offer. How could we host LOB apps like Swyx MSI, Acrobat + MST, Infor MSI + MST, etc.

**Reply by mherrmann** [reply]
@Karl-WE you can just upload them through winget.Pro's web interface or API.

**Comment by ABAG603** [comment]
Hi @denelon . Are REST source requirements documented somewhere (all required endpoints, requests and responses)?

**Reply by ABAG603** [reply]
Is this is the one to be used https://github.com/microsoft/winget-cli-restsource/blob/main/documentation/WinGet-1.9.0.yaml or more docs exits?

**Reply by vedantmgoyal9** [reply]
@ABAG603 that is the one to be used. fwiw you can paste its contents in the editor at https://editor-next.swagger.io to visualize it.

**Comment by byjrack** [comment]
I know when I have tried to chase this one it really felt like the easiest option was actually to do the sqlite source.  One of those projects I pick up and put down as i trip over "internal" sources.  Running a full REST stack just felt heavy so it became sort of a game to see what i could pull out of the sqlite db without causing winget to see the source as corrupt.  So I could use a repo and a CI job to assemble my own source db that i could publish as static content.  Our custom manifest list is a rounding error of the community registry so some of the methods like pathparts to reduce duplication and db bloat are less critical.  Still hoping we will get a source better aligned with static content at a reasonable scale at some point.  Repo or a bucket point it at a manifest in some format and call it a day.  Regrettably my cpp skills are long gone so doing a PR is a bit beyond my skills.

**Reply by florelis** [reply]
This is not officially supported or documented (I'm not even sure why) but...

If you're really inclined to create your own pre-indexed source, you could do worse than looking at the WinGetUtil. It has [C headers](https://github.com/microsoft/winget-cli/tree/master/src/WinGetUtil/WinGetUtil.h) and a [C# wrapper](https://github.com/microsoft/winget-cli/tree/master/src/WinGetUtilInterop/Interfaces), and it is published as a [NuGet package](https://www.nuget.org/packages/Microsoft.WindowsPackageManager.Utils)

You could also take inspiration from how we [set up a source for E2E tests](https://github.com/microsoft/winget-cli/tree/master/src/AppInstallerCLIE2ETests#wingetsourcecreator) which is also a good reference of [how to use the WinGetUtil API](https://github.com/microsoft/winget-cli/tree/master/src/WinGetSourceCreator/WinGetLocalSource.cs)

Again, this isn't officially supported AFAIK, so use at your own risk 🤷‍♀️

**Reply by byjrack** [reply]
Having to reverse engineer the index.db to make a guess at what was important wasn't too hard.  Was one of those random lab experiments to help divine how things actually worked.  And it functions in my case (for now) but to the supported/documented I am still hoping we can get an official source type better suited for static hosting.  For now I can loop over a folder of manifests and create an index.db that seems to meet the need, but it's far from a mission critical service on my side.  Best corollary was the homebrew tap model which i think was in one of the posts from a couple years back.  Its a good point though I should probably look at Util versus my hack job. 

I am still hoping the Intune team would support allowing community-pkgs as a built in feature instead of only windows store.  Lots of the context management, remote orchestration, UI, updates, etc. would be done for orgs that wanted to implement it.  Again we worked around the limits with scripted orchestration, but 10x harder than having it added to agentexecutor.

**Comment by dev-fYnn** [comment]
You could try https://winget-repo.io/ . It's a private and open-sourceWinget repository with some unique features. It can be deployed locally in your company or as cloud.

**Reply by byjrack** [reply]
Yeah AGPL has some sharper edges that I would prefer to avoid combined with a relatively heavy backend.  Gut says most folks are more interested in having a pure static option.  If you are running hundreds of instances than the cost may be worth it, but if you are dealing w a handful the value isn't there.  

Given the integration between winget and Intune for the Store support I am hoping that we will get more in that space in any case.  Allow folks to source Company Portal scoped installs via winget instead of requiring the UI.
